weblate/api/serializers.py

```diff
@@ -478,6 +487,52 @@ def to_internal_value(self, data):
             result["source_language"] = Language.objects.get(
                 code=result["source_language"]["code"]
             )
+        if "docfile" in result:
+            # Create fake component (needed to calculate path)
+            fake = Component(
+                project=result["project"],
+                slug=result["slug"],
+                name=result["name"],
+            )
+
+            # Create repository
+            uploaded = result["docfile"]
+            ext = os.path.splitext(os.path.basename(uploaded.name))[1]
+            format_cls = FILE_FORMATS[result["file_format"]]
+            filename = "{}/{}{}".format(
+                result["slug"],
+                result["source_language"]["code"]
+                if "source_language" in result
+                else "en",
+                ext,
+            )
+            LocalRepository.from_files(
+                fake.full_path,
+                {
+                    filename: uploaded.read(),
+                    result["template"]: format_cls.get_new_file_content(),
+                },
+            )
+            result.pop("docfile")
+        if "zipfile" in result:
+            # Create fake component (needed to calculate path)
+            fake = Component(
+                project=result["project"],
+                slug=result["slug"],
+                name=result["name"],
+            )
+            format_cls = FILE_FORMATS[result["file_format"]]
+
+            # Create repository
+            try:
+                LocalRepository.from_zip(fake.full_path, result["zipfile"])
+                LocalRepository.from_files(
+                    fake.full_path,
+                    {result["template"]: format_cls.get_new_file_content()},
+                )
+            except BadZipfile:
+                raise serializers.ValidationError("Failed to parse uploaded ZIP file.")
+            result.pop("zipfile")
```

Comment:
This all duplicates existing code from `weblate/trans/views/create.py`. It should be extracted from there and reused in the API. It might to to the `weblate/trans/models/component.py` (where component related code is) or `weblate/utils/views.py` (where other view helpers are already stored).